In [1]:
import arviz as az
import numpy as np
import pymc3 as pm

# Data Extraction

In [2]:
data_array = np.loadtxt('bladderc.csv', delimiter=',', skiprows=1)
time_data = data_array[:, 0]        # time at which the observation was made
observed_data = data_array[:, 1]    # whether or not cancer was observed
group_data = data_array[:, 2]       # whehter or not placebo or treatment group


# Separate data into censored and uncensored observations
observed_mask = observed_data == 1
censored = observed_data[~observed_mask]

time_cens = time_data[~observed_mask]
time_uncens = time_data[observed_mask]
group_cens = group_data[~observed_mask]
group_uncens = group_data[observed_mask]    

# Model

In [5]:
with pm.Model() as m:
    # Priors
    beta0 = pm.Normal('beta0', mu=0, tau=0.0001)
    beta1 = pm.Normal('beta1', mu=0, tau=0.0001)

    # Rate parameters
    λ_cens = pm.math.exp(beta0 + beta1 * group_cens)
    λ_uncens = pm.math.exp(beta0 + beta1 * group_uncens)

    # Impute censored data
    BoundedExponential = pm.Bound(pm.Exponential, lower=np.min(censored))
    impute_censored = BoundedExponential('impute_censored', lam=λ_cens, shape=censored.shape[0])

    # Likelihood
    likelihood = pm.Exponential(
        "likelihood",
        lam=λ_uncens,
        observed=time_uncens,
        shape=time_uncens.shape[0],
    )

    # Means for placebo and treatment times
    μ0 = pm.Deterministic('μ0', pm.math.exp(-beta0))
    μ1 = pm.Deterministic('μ1', pm.math.exp(-(beta0 + beta1)))

    # Difference of means
    diff_means = pm.Deterministic('diff_means', μ1 - μ0)

    # Hypothesis testing
    hypothesis = pm.Deterministic("hypothesis", pm.math.switch(pm.math.ge(μ1, μ0), 1, 0))

    trace = pm.sample(5000, target_accept=0.95)

/Users/adamcuculich/georgia-tech/bayesian-stats/env/lib/python3.9/site-packages/deprecat/classic.py:215: FutureWarning: In v4.0, pm.sample will return an `arviz.InferenceData` object instead of a `MultiTrace` by default. You can pass return_inferencedata=True or return_inferencedata=False to be safe and silence this warning.
  return wrapped_(*args_, **kwargs_)
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [impute_censored, beta1, beta0]


/Users/adamcuculich/georgia-tech/bayesian-stats/env/lib/python3.9/site-packages/scipy/stats/_continuous_distns.py:624: RuntimeWarning: overflow encountered in _beta_ppf
  return _boost._beta_ppf(q, a, b)
/Users/adamcuculich/georgia-tech/bayesian-stats/env/lib/python3.9/site-packages/scipy/stats/_continuous_distns.py:624: RuntimeWarning: overflow encountered in _beta_ppf
  return _boost._beta_ppf(q, a, b)
/Users/adamcuculich/georgia-tech/bayesian-stats/env/lib/python3.9/site-packages/scipy/stats/_continuous_distns.py:624: RuntimeWarning: overflow encountered in _beta_ppf
  return _boost._beta_ppf(q, a, b)
/Users/adamcuculich/georgia-tech/bayesian-stats/env/lib/python3.9/site-packages/scipy/stats/_continuous_distns.py:624: RuntimeWarning: overflow encountered in _beta_ppf
  return _boost._beta_ppf(q, a, b)
Sampling 4 chains for 1_000 tune and 5_000 draw iterations (4_000 + 20_000 draws total) took 16 seconds.


# Show Statistics

In [6]:
az.summary(trace)

Got error No model on context stack. trying to find log_likelihood in translation.
/Users/adamcuculich/georgia-tech/bayesian-stats/env/lib/python3.9/site-packages/arviz/data/io_pymc3_3x.py:98: FutureWarning: Using `from_pymc3` without the model will be deprecated in a future release. Not using the model will return less accurate and less useful results. Make sure you use the model argument or call from_pymc3 within a model context.
  warnings.warn(


mean      sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  \
beta0                -2.279   0.185  -2.634   -1.943      0.002    0.001   
beta1                -0.064   0.300  -0.628    0.499      0.003    0.002   
impute_censored[0]   10.005  10.377   0.000   28.995      0.067    0.051   
impute_censored[1]    9.812  10.238   0.000   28.067      0.068    0.050   
impute_censored[2]    9.913  10.282   0.001   28.418      0.065    0.051   
impute_censored[3]    9.988  10.346   0.002   28.492      0.069    0.054   
impute_censored[4]   10.015  10.288   0.002   28.204      0.064    0.050   
impute_censored[5]    9.817  10.213   0.000   28.293      0.060    0.049   
impute_censored[6]    9.900  10.333   0.001   28.576      0.067    0.051   
impute_censored[7]    9.929  10.249   0.004   28.248      0.065    0.050   
impute_censored[8]    9.869  10.184   0.000   27.975      0.063    0.049   
impute_censored[9]    9.817  10.111   0.001   28.197      0.061    0.049   
impute_censored[10]   9.994  10.258   0.001   28.347      0.064    0.051   
impute_censored[11]   9.958  10.399   0.000   28.485      0.065    0.053   
impute_censored[12]   9.955  10.328   0.000   28.408      0.063    0.052   
impute_censored[13]   9.909  10.213   0.006   28.047      0.063    0.051   
impute_censored[14]   9.870  10.170   0.003   28.190      0.065    0.051   
impute_censored[15]   9.914  10.248   0.002   28.052      0.067    0.051   
impute_censored[16]  10.020  10.350   0.000   28.518      0.063    0.052   
impute_censored[17]   9.902  10.107   0.000   28.079      0.065    0.053   
impute_censored[18]   9.863  10.095   0.001   28.184      0.062    0.051   
impute_censored[19]  10.677  11.238   0.001   30.372      0.070    0.059   
impute_censored[20]  10.726  11.595   0.001   31.526      0.074    0.060   
impute_censored[21]  10.608  11.240   0.001   30.179      0.073    0.059   
impute_censored[22]  10.666  11.366   0.001   30.930      0.076    0.058   
impute_censored[23]  10.819  11.495   0.002   31.027      0.072    0.057   
impute_censored[24]  10.739  11.576   0.000   31.577      0.074    0.060   
impute_censored[25]  10.511  11.056   0.001   29.705      0.071    0.055   
impute_censored[26]  10.736  11.279   0.000   30.841      0.073    0.056   
impute_censored[27]  10.710  11.494   0.001   31.232      0.072    0.058   
impute_censored[28]  10.904  11.615   0.001   31.101      0.073    0.059   
impute_censored[29]  10.666  11.269   0.000   30.651      0.071    0.056   
impute_censored[30]  10.639  11.261   0.000   30.331      0.074    0.057   
impute_censored[31]  10.617  11.389   0.000   30.811      0.071    0.057   
impute_censored[32]  10.830  11.676   0.001   31.585      0.073    0.059   
impute_censored[33]  10.729  11.565   0.000   31.359      0.072    0.062   
impute_censored[34]  10.652  11.321   0.000   30.486      0.069    0.056   
impute_censored[35]  10.757  11.489   0.001   31.236      0.071    0.060   
impute_censored[36]  10.611  11.114   0.000   29.852      0.069    0.055   
impute_censored[37]  10.579  11.117   0.001   30.295      0.074    0.056   
impute_censored[38]  10.685  11.267   0.000   30.415      0.074    0.057   
μ0                    9.934   1.882   6.712   13.496      0.019    0.013   
μ1                   10.709   2.644   6.397   15.721      0.023    0.016   
diff_means            0.775   3.238  -5.347    6.873      0.033    0.023   
hypothesis            0.580   0.494   0.000    1.000      0.005    0.003   

                     ess_bulk  ess_tail  r_hat  
beta0                 10557.0   13425.0    1.0  
beta1                  9795.0   12705.0    1.0  
impute_censored[0]    15380.0    9188.0    1.0  
impute_censored[1]    13314.0    7989.0    1.0  
impute_censored[2]    15791.0    8879.0    1.0  
impute_censored[3]    15059.0    9089.0    1.0  
impute_censored[4]    17154.0    8793.0    1.0  
impute_censored[5]    17905.0    8946.0    1.0  
impute_censored[6]    15606.0    9159.0    1.0  
impute_censored[7]    16435.0    979

# Conclusion

### Is the 95% Credible Set for μ1 − μ0 all positive?
 The 95% credible set ranges from approximately -5.347 to 6.873. Because this interval includes negative values, it is not all positive. Therefore, based on this credible set alone, we cannot confidently state that the treatment increases the time to cancer recurrence compared to the placebo.

 ### What is the posterior probability of hypothesis H : μ1 > μ0?
 The mean of the hypothesis variable in the trace (approximately 0.58) represents the proportion of the posterior distribution where μ1 is greater than μ0. Based on this, there is an estimated 58% posterior probability that the treatment group has a longer time to cancer recurrence than the placebo group.

 ### Comment on the benefits of the treatment.
 Based on the results, there is evidence that suggests the treatment may increase the time to cancer recurrence, but the evidence is not strong. The 95% credible set for the difference in means includes zero, which indicates that we cannot be 95% confident that the treatment provides a benefit over placebo in terms of increasing the time to cancer recurrence. However, the posterior probability that the treatment is better than the placebo is 58%, which indicates that there is some evidence in favor of the treatment's efficacy--that said, it is not conclusive. Further analysis or data may be required to make a stronger statement regarding the benefits of the treatment.